src/index.js

```diff
@@ -51,20 +51,30 @@ module.exports = {
   },
 
   async: function() {
-    return {
-      get: async () => {
+    return new Proxy({
+      get: async (key) => {
         await this.initialize();
         const value = _.get(this, key, null);
         if (value) {
           return value;
         }
-        return await Config.get(key);
+        return await this.config.get(key);
+      },
+      submodule: (name) => {
+        return this.submodule(name);
+      }
+    }, {
+      get: function(obj, prop) {
+        if (prop in obj) {
+          return obj[prop];
+        }
+        return obj.get(prop);
       }
-    };
+    });
```

Comment:
The Proxy API is super cool, but what makes it a better choice here than an object with `get` and `submodule` methods? It doesn't seem like it simplifies the interface for the consumer; they'll need to know that accessing config-specific properties returns a Promise, whereas accessing properties that they themselves stored on the proxy returns the actual value. That seems less clear to me than having a specific method for async retrieval. 

On the other hand, it just adds some weirdness which doesn't make sense unless you know what's going on under the hood:

```
let config = Common.async();
'Logger' in config // false
config.Logger === config.Logger // false
```

Reply:
I went back and forth a bit in my head on whether or not this was an OK approach.  I certainly don't feel like "await" plus a property is terribly intuitive.  However, I found that in some cases I was writing a bunch of code that basically fit the form "config.get('something')" over and over and over again (especially in lambdas that use a bunch of what c-l-c vends) and it just felt super verbose, and under the hood what we are implementing here is basically a dictionary anyway (albeit potentially lazily loaded).  So why not make it look and operate like a dictionary instead of some weird "like a dictionary but you get at things with a 'get' method instead of the standard way".

I think having this all behind async() or legacy() does communicate some of the expectations of the API, so I don't think this would necessarily be a total surprise to a caller.  But yeah it is a bit weird.  If folks are bugged by this I can pull the proxy stuff out or think there's a high probability of misunderstanding / misuse, I'm OK going back to the wordier approach.  I just thought this might make for some cleaner code inside the actual lambdas.

Reply:
It's up to you. I personally feel like this is pushing the limit in terms of predictable getter behavior (especially since it could technically trigger initialization of the entire system, including multiple external requests, albeit only once). At the same time, the impact here is pretty contained. We can roll with it and see if anyone implementing it pipes up.